File: test/conformance/js/rows/C9.mjs
// THE UNION TAG PAST ARMS → NONE ASSERTION (SPEC §3.4, FIXED-FORM-ALGORITHM.md §4.6)
//
// The Effect union in WeaponConfig has two arms: Buff(1), Debuff(2).
// Tag 0 = None, tags > 2 are past the arm count and MUST land as None (0)
// with report.clamped incremented by exactly 1.
//
// DERIVATION OF THE WIRE VECTOR:
//   WeaponConfig body layout (fixed-form, little-endian):
//     offset 0: damage      float32 (4 bytes)
//     offset 4: speed       float32 (4 bytes)
//     offset 8: penetration int32   (4 bytes)
//     offset 12: channel    bits(6) stored as uint32 (4 bytes)
//     offset 16: homing     bool (1 byte)
//     offset 17: Effect.Type tag (1 byte) — FORGED to 9 here, past arm count 2
//     offset 18: payload    float32 (4 bytes, Buff.Debuff.Multiplier or Debuff.Amount)
//   Total body size: 22 bytes. No length/count/terminator — fixed form is positional only.
//
// This is a CONFORMANCE TEST for cell js/C9 in the schema matrix.
// It uses the same generated JS tables code path that the driver imports from
// build/tables-generated-js/, directly calling the decode function.

import { fileURLToPath } from "node:url";
import { join, dirname, resolve } from "node:path";
import { pathToFileURL } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, "../../../..");
const generated = process.env.SCHEMA_JS_GENERATED ?? join(repoRoot, "build/tables-generated-js");
const mod = await import(pathToFileURL(join(generated, "examples/TabledemoTable.js")).href);

const { WeaponConfig, InitWeaponConfig, WeaponConfigFixedDecode, TableFixedReport } = mod;

let failures = 0;
function assert(condition, msg) {
  if (!condition) {
    console.log(`FAIL: ${msg}`);
    failures++;
  } else {
    console.log(`PASS: ${msg}`);
  }
}

// Construct the 22-byte body vector with EFFECT TAG = 9 (past arms count of 2).
// The needle value 0x09 at offset 17 names no arm: the reader knows none, so
// it must refuse-to-decode this arm and fall back to None.
function makeBody(tag) {
  const buf = new ArrayBuffer(22);
  const view = new DataView(buf);
  view.setFloat32(0, 21.0, true);   // damage (any valid value)
  view.setFloat32(4, 500.0, true);  // speed
  view.setInt32(8, 5, true);        // penetration
  view.setUint32(12, 2, true);      // channel
  view.setUint8(16, 1);             // homing = true
  view.setUint8(17, tag);           // Effect.Type — FORGED VALUE
  view.setFloat32(18, 1.5, true);   // payload
  return new Uint8Array(buf);
}

// ---- THE LAW: union tag past arms → None, clamped == 1 ----

const body = makeBody(9); // tag = 9, well past arm count of 2

const wc = new WeaponConfig();
InitWeaponConfig(wc);
const report = new TableFixedReport();

WeaponConfigFixedDecode(wc, new DataView(body.buffer), 0, report);

assert(wc.Effect.Type === 0, `tag 9 lands as None (Type=${wc.Effect.Type})`);
assert(report.clamped === 1, `report.clamped is exactly 1 (got ${report.clamped})`);
assert(report.refused === 0, `not a refusal (refused=${report.refused})`);
assert(!report.malformed, "not malformed");
assert(report.unknown === 0, "unknown = 0");
assert(report.kindMismatch === 0, "kindMismatch = 0");
assert(report.widened === 0, "widened = 0");
assert(report.duplicate === 0, "duplicate = 0");

// Verify neighbouring scalar after the union field is untouched.
// Penetration starts at 5 (valid range [0,10]) and should be preserved.
assert(wc.Penetration === 5, `neighbour penetration = 5 (${wc.Penetration})`);

console.log(failures === 0 ? "GREEN" : "RED");
process.exit(failures > 0 ? 1 : 0);

File: test/conformance/c/rows/C9.c
/* cell c/C9: union tag past arms -> None
   law: FIXED-FORM-ALGORITHM.md:945, SPEC-TABLES.md:6736
   production entrypoint: V1Table.h:6873 schema_tblv1_cfg_fixed_clamp_body_
   call site: V1Table.h:7524 cfg_fixed_load -> schema_tblv1_cfg_fixed_clamp_ -> clamp_body_
   writes a Cfg record with effect.type=WARD, corrupts the tag to 3 (past EFFECT_TYPE_MAX=2),
   loads through cfg_fixed_load, asserts the tag clamps to EFFECT_TYPE_NONE and
   the report counts one clamped event. */

#include "V1Table.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(void)
{
    int failures = 0;

    int64_t need = cfg_fixed_measure(1);

    uint8_t * buf = (uint8_t *) malloc((size_t) need);
    if (!buf) { fprintf(stderr, "FAIL: malloc\n"); return 1; }

    Cfg value;
    memset(&value, 0, sizeof(value));
    cfg_reset(&value);
    value.effect.type = EFFECT_TYPE_WARD;
    value.effect.as.ward.charge = 3.14f;

    int64_t written = cfg_fixed_save(&value, 1, buf, need);
    if (written < 0) { fprintf(stderr, "FAIL: cfg_fixed_save\n"); free(buf); return 1; }

    /* The record body starts at body_offset + 8 (after the per-record hash).
       Effect.type is at body byte 137 (V1Table.h:6763). */
    int64_t body_offset = kTableFixedHeaderBytes + 4 + (int64_t) sizeof(cfg_fixed_layout);
    int64_t record_start = body_offset; /* start of this record (includes hash) */
    int64_t tag_offset = record_start + 8 + 137; /* skip hash, then body[137] = Effect.type */

    /* Verify the tag is WARD(2) before corruption */
    if (buf[tag_offset] != 2) { fprintf(stderr, "BUG: expected tag=2 before corruption, got %d\n", buf[tag_offset]); free(buf); return 1; }

    /* Corrupt the tag to 3 (past EFFECT_TYPE_MAX=2) */
    buf[tag_offset] = 3;

    /* Load it back */
    Cfg loaded;
    memset(&loaded, 0, sizeof(loaded));
    TableReport report;
    memset(&report, 0, sizeof(report));

    int64_t n = cfg_fixed_load(&loaded, 1, buf, written, NULL, 0, NULL, &report);
    if (n != 1) { fprintf(stderr, "FAIL: cfg_fixed_load returned %ld\n", (long)n); free(buf); return 1; }

    /* Assert: tag past max -> None */
    if (loaded.effect.type != EFFECT_TYPE_NONE) {
        fprintf(stderr, "FAIL: union tag past arms -> None: expected type=%d, got %d\n",
                EFFECT_TYPE_NONE, loaded.effect.type);
        failures++;
    } else {
        printf("PASS: union tag past arms -> None (tag 3 -> EFFECT_TYPE_NONE)\n");
    }

    /* Assert: clamped counter incremented */
    if (report.clamped < 1) {
        fprintf(stderr, "FAIL: expected clamped >= 1, got %d\n", report.clamped);
        failures++;
    } else {
        printf("PASS: clamped counter = %d\n", report.clamped);
    }

    /* Assert: arm payload is undisturbed (None arm has no payload) */
    if (loaded.effect.as.boost.power != 0) {
        fprintf(stderr, "FAIL: None arm payload should be zero, got %d\n",
                loaded.effect.as.boost.power);
        failures++;
    } else {
        printf("PASS: None arm payload is zero\n");
    }

    free(buf);

    if (failures) {
        fprintf(stderr, "FAIL: %d assertion(s) failed\n", failures);
        return 1;
    }
    return 0;
}